src/v/cloud_roles/request_response_helpers.cc

```diff
@@ -43,46 +44,36 @@ ss::future<api_response> make_request(
   http::client client,
   http::client::request_header req,
   std::chrono::milliseconds timeout) {
-    try {
-        auto response_stream = co_await client.request(std::move(req), timeout);
-        auto status = co_await get_status(response_stream);
-        if (
-          std::find(
-            retryable_http_status.begin(), retryable_http_status.end(), status)
-          != retryable_http_status.end()) {
-            co_return api_request_error{
-              .reason = fmt::format("http request failed:{}", status),
-              .error_kind = api_request_error_kind::failed_retryable};
-        }
-
-        if (status != boost::beast::http::status::ok) {
-            co_return api_request_error{
-              .reason = fmt::format("http request failed:{}", status),
-              .error_kind = api_request_error_kind::failed_abort};
-        }
-
-        auto data = co_await drain_response_stream(std::move(response_stream));
-        co_await client.stop();
-        client.shutdown();
-        co_return data;
-    } catch (const std::system_error& ec) {
-        if (auto code = ec.code(); std::find(
-                                     retryable_system_error_codes.begin(),
-                                     retryable_system_error_codes.end(),
-                                     code.value())
-                                   == retryable_system_error_codes.end()) {
-            co_return api_request_error{
-              .reason = ec.what(),
-              .error_kind = api_request_error_kind::failed_abort};
-        }
-        co_return api_request_error{
-          .reason = ec.what(),
-          .error_kind = api_request_error_kind::failed_retryable};
-    } catch (const std::exception& e) {
-        co_return api_request_error{
-          .reason = e.what(),
-          .error_kind = api_request_error_kind::failed_abort};
-    }
+    co_return co_await http::with_client(
+      std::move(client),
+      [req = std::move(req),
+       &timeout](auto& client) mutable -> ss::future<api_response> {
+          co_return co_await do_make_request(
+            req,
+            [&client, &timeout](auto& req) mutable -> ss::future<api_response> {
```

Comment:
do you think that the lifetime of these captures are going to be ok and not subject to the issues with coroutine-lambdas? since you are co-awaiting all of these, perhaps you could use https://github.com/scylladb/seastar/blob/master/doc/lambda-coroutine-fiasco.md#solution ?

Reply:
I think in this case it may be okay (please correct me if not), because the captures by references we have here are `client` and `timeout`, client is wrapped in a `do_with` by `http::with_client` so it should be on the heap until the `do_with` completes, and `timeout` is not used past the first co_await. timeout could also just be a copy.

I do think the problem will be in the second change in this pr, in `post_request` where we capture `stream` by reference and access it after the first `co_await`, I will change it to use `coroutine::lambda`.